panc/src/main/scripts/panlint

```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/env python
+# coding=utf8
+
+import re
+import argparse
+from glob import glob
+from colorama import Fore, Style, init as colorama_init
+from sys import stdout, exit as sys_exit
+from inspect import getmembers, ismethod
+from prettytable import PrettyTable
+
+RS_COMMENT = r'(?:#.*|@{.*})$'
+
+RE_STRING = re.compile(r'''('.*?'|".*?"(?<!\\"))''')
+
+RE_FIRST_LINE = re.compile(r'^(?:(?:declaration|unique|structure|object) )?template (?:(?:[\w-]+/)+)?[\w-]+;$')
+RE_COMMENT_LINE = re.compile(r'^\s*' + RS_COMMENT)
+RE_TRAILING_COMMENT = re.compile(r'\s*' + RS_COMMENT)
+RE_ANNOTATION = re.compile(r'@{(?:.|\n)*?}')
+
+# Simple regular-expression based checks that will be performed against all non-ignored lines
+# Every pattern must provide a single capturing group named "error"
+LINE_PATTERNS = {
+    "Indentation should be a multiple of four spaces" : re.compile(r'^(?!(    )*(\S|$))(?P<error>\s*)'),
+    "Spaces should be used instead of tabs" : re.compile(r'(?P<error>\t+)'),
+    "Trailing whitespace" : re.compile(r'(?P<error>\s+$)'),
+    "Use dict() instead of nlist()" : re.compile(r'(?P<error>nlist)'),
+    "Include statements no longer need curly braces" : re.compile(r'''include\s+(?P<error>{.+})'''),
+    "Line is longer than 159 characters" : re.compile(r'''^.{0,159}(?P<error>.*?)$'''),
+}
+
+DEBUG = False
+
+class LineChecks:
+    """More complex single line checks that require some logic to implement their checks"""
+
+    RE_OPERATOR = re.compile(r'(.)([>=<!?]=|[+*=/-])(.)')
+
+    def whitespace_around_operators(self, line, string_ranges):
+        """Check a line of text to ensure that there is whitespace before and after all operators"""
+        operators = self.RE_OPERATOR.finditer(line)
+
+        passed = True
+        message = set()
+
+        diagnosis = ' ' * len(line)
+
+        for operator in operators:
+            char_before, _, char_after = operator.groups()
+            s, e = operator.span(2)
+
+            if not inside_string(s, e, string_ranges):
+                if char_before not in (' ', '\t'):
+                    passed = False
+                    message.add('before')
+                    s -= 1
+                if char_after not in (' ', '\t'):
+                    passed = False
+                    message.add('after')
+                    e += 1
+
+                if not passed:
+                    diagnosis = diagnosis[:s] + ('^' * (e-s)) + diagnosis[e:]
+
+        message = 'Missing space %s operator' % ' and '.join(message)
+        return (passed, diagnosis, message)
+
+
+def inside_string(i, j, string_ranges):
+    """Returns true if the range described by i and j is contained within a range in the list string_ranges"""
+    for s, e in string_ranges:
+        if i >= s and j <= e:
+            return True
+    return False
+
+
+def print_fileinfo(filename, line_number, message, vi=False):
+    """Returns a formatted string with filename, line_number and message"""
```

Comment:
It outputs line numbers as `+nnn` and prefixes descriptive text with a `#` so that each line can be passed to vi. It's the `--vi` option being rather inelegantly handled.